### PR TITLE
HIVE-28428 Map hash aggregation performance degradation

### PR DESCRIPTION
**Pull request by ryukobayashi**
### What changes were proposed in this pull request?

Fix performance regression in map-hash aggregation. The following log will be output in large volume, causing performance degradation.
https://github.com/apache/hive/blob/rel/release-4.0.0/ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java#L967
https://github.com/apache/hive/blob/rel/release-4.0.0/ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java#L978

And, currently the flush size is fixed, but performance can be improved by changing it depending on the data

### Why are the changes needed?

Prevents performance degradation.

### Is the change a dependency upgrade?

Yes, we can improve performance by changing the flash size depending on your data.

### How was this patch tested?

Performance was checked in a local environment.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java [hunk 1]
```diff
@@ -964,7 +963,8 @@ private void flushHashTable(boolean complete) throws HiveException {
     }
 
     int oldSize = hashAggregations.size();
-    LOG.info("Hash Tbl flush: #hash table = {}", oldSize);
+    int flushSize = (int) (oldSize * hashAggrFlushPercent);
+    LOG.trace("Hash Tbl flush: #hash table = {}", oldSize);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java:967]: should we log the flush size as well?

**Comment by deniskuzZ** [comment]
@ryukobayashi is it possible to create some test for that? maybe something similar to `tpcds30tb` tests

File: ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java (hunk 1 above)
Comment (ryukobayashi) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java:967]: OK, I will log it.

**Comment by ryukobayashi** [comment]
@deniskuzZ Since I'm not familiar with `tpcds30tb`, I tried adding a test in the same way. Please let me know if there's anything wrong.

**Comment by deniskuzZ** [comment]
> @deniskuzZ Since I'm not familiar with `tpcds30tb`, I tried adding a test in the same way. Please let me know if there's anything wrong.

yes, it was added in the right place. tpcds30tb doesn't have the 30tb data but it simulates it via stats. Does it cover you test scenario, is the output expected?

**Comment by ryukobayashi** [comment]
@deniskuzZ Yes, I have created and checked a test scenario for this issue.

**Comment by deniskuzZ** [comment]
> @deniskuzZ Yes, I have created and checked a test scenario for this issue.

how so, I don't see you override `hive.map.aggr.hash.fulush.size.percent` in a test, so it's using the default 0.1

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 2]
```diff
@@ -2012,6 +2012,9 @@ public static enum ConfVars {
         "Set to 1 to make sure hash aggregation is never turned off."),
     HIVE_MAP_AGGR_HASH_MIN_REDUCTION_LOWER_BOUND("hive.map.aggr.hash.min.reduction.lower.bound", (float) 0.4,
         "Lower bound of Hash aggregate reduction filter. See also: hive.map.aggr.hash.min.reduction"),
+    HIVE_MAP_AGGR_HASH_FLUSH_SIZE_PERCENT("hive.map.aggr.hash.fulush.size.percent", (float) 0.1,
```
[the remark this answers is not in the document]
Comment (ryukobayashi) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: fixed.

**Comment by ryukobayashi** [comment]
@deniskuzZ Right, it does not directly affect the plan.

**Review by deniskuzZ — APPROVED** [review]
LGTM +1

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 3]
```diff
@@ -2012,6 +2012,9 @@ public static enum ConfVars {
         "Set to 1 to make sure hash aggregation is never turned off."),
     HIVE_MAP_AGGR_HASH_MIN_REDUCTION_LOWER_BOUND("hive.map.aggr.hash.min.reduction.lower.bound", (float) 0.4,
         "Lower bound of Hash aggregate reduction filter. See also: hive.map.aggr.hash.min.reduction"),
+    HIVE_MAP_AGGR_HASH_FLUSH_SIZE_PERCENT("hive.map.aggr.hash.flush.size.percent", (float) 0.1,
+        "Whether percentage of the map side group aggregation hash tables to flush.　\n" +
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2016]: @ryukobayashi, could you please update the description before the merge
`Percentage of entries to flush in the map-side group aggregation hash table`

**Comment by deniskuzZ** [comment]
> @deniskuzZ Right, it does not directly affect the plan.

in that case, should we keep the added q test?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 3 above)
Comment (ryukobayashi) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2016]: Got it. will do.

**Comment by ryukobayashi** [comment]
@deniskuzZ I think it is not necessary so I will remove it.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 4]
```diff
@@ -2012,6 +2012,9 @@ public static enum ConfVars {
         "Set to 1 to make sure hash aggregation is never turned off."),
     HIVE_MAP_AGGR_HASH_MIN_REDUCTION_LOWER_BOUND("hive.map.aggr.hash.min.reduction.lower.bound", (float) 0.4,
         "Lower bound of Hash aggregate reduction filter. See also: hive.map.aggr.hash.min.reduction"),
+    HIVE_MAP_AGGR_HASH_FLUSH_SIZE_PERCENT("hive.map.aggr.hash.flush.size.percent", (float) 0.1,
+        "Percentage of hash table entries to flush in map-side group aggregation.\n" +
+        "Flushes the specified percentage in map-side group aggregation hash table."),
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2017]: @ryukobayashi, please remove that, it just repeats the above statement

Comment (ryukobayashi) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: Oh OK, done.

**Pull request merged**